### PR TITLE
Feature/inventory saga implementation

### PR DESCRIPTION
**Pull request by KiLanger**
Issue: fde091db-dcff-4602-8531-a39a05475728
- [x]  Requirements of the issue are met
- [x]  Code has been reviewed
- [x]  Code is documented
- [x]  GraphQL related artefacts are documented

### CODE REVIEW

File: src/events/dto/create-order.dto.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    @IsInt()
+    @Min(0)
+    compensatableAmount: number;
+
+    @ValidateNested()
+    @Type(() => ShipmentDTO)
+    shipment: ShipmentDTO;
+
+    @IsArray()
+    @IsUUID("4", { each: true })
```
Comment (nk-coding) [inline src/events/dto/create-order.dto.ts:93]: why require version 4 here but nowhere else?
or is it required to vige a version when you gibe the options object?

File: src/events/event.controller.ts [hunk 2]
```diff
@@ -1,29 +1,119 @@
...
+  async subscribeToProductVariantEvent(@Body() body: any): Promise<void> {
     // Handle incoming event data from Dapr
     const id = body.data.id;
     this.logger.log(`Received event for product variant with id: ${id}`);
     // Call the product variant service to create a new product variant
     this.productVariantService.create(id);
   }
+
+  @Post('order-created')
+  async subscribeToOrderEvent(@Body('data') orderDto: CreateOrderDto): Promise<void> {
```
Comment (nk-coding) [inline src/events/event.controller.ts:43]: this exceeds the 20 LOC rule by Becker a lot, maybe extracting some code?

File: src/inventory.gql [hunk 3]
```diff
@@ -27,6 +27,9 @@ type ProductItem
 
   """Describes the inventory status of the product item"""
   inventoryStatus: ProductItemStatus!
+
+  """Reference to the order that reserved the product item"""
+  orderId: UUID!
```
Comment (nk-coding) [inline src/inventory.gql:32]: shouldn't this be nullable?

**Review by nk-coding — CHANGES_REQUESTED** [review]
So, basically nothing here has jsdoc comments.
I know you want to do this in a future PR, the problem is, we already got some heat from supervisors for this, so I think it would be a good start to already add jsdoc here to the new stuff (a bit as a sign of good will to the supervisors)

File: src/events/dto/create-order.dto.ts (hunk 1 above)
Comment (KiLanger) [reply src/events/dto/create-order.dto.ts:93]: Yes its necessary when using options

File: src/events/dto/create-order.dto.ts [hunk 4]
```diff
@@ -1,32 +1,49 @@
...
   @IsUUID()
   shipmentMethodId: string;
+
+
 }
 
+/**
+ * DTO representing an order.
+ */
 export class OrderDTO {
```
Comment (nk-coding) [inline src/events/dto/create-order.dto.ts:47]: all the properties also need docs

File: src/events/event.controller.ts [hunk 5]
```diff
@@ -16,6 +16,11 @@ export class EventController {
   ) {}
 
   @Get('/dapr/subscribe')
+  /**
```
Comment (nk-coding) [inline src/events/event.controller.ts:19]: docs should be placed above the decorator

File: src/events/event.controller.ts [hunk 6]
```diff
@@ -86,12 +87,43 @@ export class EventController {
...
 
-  createInventorySuccessEvent(orderId: OrderDTO): void {
-    // Create an event to indicate that product items have been reserved
+  
+  /**
+   * Reserves product items in batch for an order.
+   * @param order - The order containing the order items to be reserved.
+   * @returns A promise that resolves to an array of objects containing the productVariantId and the reservation result.
+   */
+  private async batchPromiseOrderItems(order: OrderDTO) {
```
Comment (nk-coding) [inline src/events/event.controller.ts:96]: return type not defined explicitely

File: src/inventory/entities/product-item.entity.ts [hunk 7]
```diff
@@ -36,6 +36,7 @@ export class ProductItem {
 
   @Prop()
   @Field(() => UUID, {
+    nullable: true,
```
Comment (nk-coding) [inline src/inventory/entities/product-item.entity.ts:39]: did you regenerate the schema file?

**Review by nk-coding — CHANGES_REQUESTED** [review]

File: src/events/event.controller.ts (hunk 5 above)
Comment (KiLanger) [reply src/events/event.controller.ts:19]: Ok. I couldnt find any reference regarding this but understand your approach.

**Comment by nk-coding** [comment]
> Requirements of the issue are met

well, which issue?
besides that looks good

**Pull request merged**